cmd/joy/release.go

```diff
@@ -241,69 +245,145 @@ func NewReleaseRenderCmd() *cobra.Command {
+						ValueMapping: cfg.ValueMapping,
+						IO: internal.IO{
+							Out: buffer,
+							Err: cmd.ErrOrStderr(),
+							In:  cmd.InOrStdin(),
+						},
+						Helm:  helm.CLI{},
+						Color: colorEnabled,
+					},
+				}, nil
+			}
+
+			renderRef := func(ref string) (result string, err error) {
+				if ref != "" {
+					dirty, err := git.IsDirty(cfg.CatalogDir)
+					if err != nil {
+						return "", fmt.Errorf("checking if catalog is in dirty state: %w", err)
+					}
+
+					if dirty {
+						if err := git.Stash(cfg.CatalogDir); err != nil {
+							return "", fmt.Errorf("stashing catalog: %w", err)
+						}
+						defer func() {
+							if applyErr := git.StashApply(cfg.CatalogDir); err == nil && applyErr != nil {
+								err = fmt.Errorf("applying stash: %w", applyErr)
+							}
+						}()
+					}
+
+					if err := git.Checkout(cfg.CatalogDir, ref); err != nil {
+						return "", fmt.Errorf("checking out: %s: %w", ref, err)
+					}
+					defer func() {
+						if swichErr := git.SwitchBack(cfg.CatalogDir); err == nil && swichErr != nil {
+							err = fmt.Errorf("switching git back to previous branch: %w", swichErr)
+						}
+					}()
+				}
+
+				var buf bytes.Buffer
+				params, err := buildRenderParams(&buf)
+				if err != nil {
+					return "", err
+				}
+
+				if err := render.Render(cmd.Context(), params); err != nil {
+					return "", err
+				}
+
+				return buf.String(), nil
 			}
 
-			io := internal.IO{
-				Out: cmd.OutOrStdout(),
-				Err: cmd.ErrOrStderr(),
-				In:  cmd.InOrStdin(),
+			gitRefResult, err := renderRef(gitRef)
+			if err != nil {
+				return err
 			}
 
-			helmCLI := helm.CLI{IO: io}
+			if diffRef == "" {
+				_, err := fmt.Fprintln(cmd.OutOrStdout(), gitRefResult)
+				return err
+			}
 
-			params := render.RenderParams{
-				Env:     env,
-				Release: releaseName,
-				Cache: helm.ChartCache{
-					DefaultChart: cfg.DefaultChart,
-					Root:         cfg.JoyCache,
-					Puller:       helmCLI,
-				},
-				Catalog: cat,
-				CommonRenderParams: render.CommonRenderParams{
-					ValueMapping: cfg.ValueMapping,
-					IO:           io,
-					Helm:         helmCLI,
-					Color:        colorEnabled,
-				},
+			diffRefResult, err := renderRef(diffRef)
+			if err != nil {
+				return err
 			}
 
-			return render.Render(cmd.Context(), params)
+			diffFunc := func() text.DiffFunc {
+				if colorEnabled {
+					return text.DiffColorized
+				}
+				return text.Diff
+			}()
+
+			diff := diffFunc(
+				text.File{Name: cmp.Or(gitRef, "(current)"), Content: gitRefResult},
+				text.File{Name: diffRef, Content: diffRefResult},
+				diffContext,
+			)
+
+			_, err = fmt.Fprint(cmd.OutOrStdout(), diff)
+			return err
 		},
 	}
 
 	cmd.Flags().StringVar(&gitRef, "git-ref", "", "git ref to checkout before render")
+	cmd.Flags().StringVar(&diffRef, "diff-ref", "", "git ref to checkout before render")
+	cmd.Flags().IntVarP(&diffContext, "diff-context", "c", 8, "line context when rendering diff")
```

Comment:
I would add an option to get a [unified diff](https://github.com/pmezard/go-difflib/blob/v1.0.0/difflib/difflib.go#L635) ouf of it. It can then be used in markdown. It can also be piped to other tools like [delta](https://github.com/dandavison/delta).

```diff
diff --git a/Makefile b/Makefile
index d7c91f3e..f33b4321 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,10 @@
-.PHONY: fmt
+.PHONY: ci fmt
 
 PRETTIER_VERSION = 3.2.5
+PRETTIER_PATTERN = '**/*.{yaml,yml,md}'
 
+ci:
+	joy release validate
+	npx prettier@'$(PRETTIER_VERSION)' $(PRETTIER_PATTERN) --check
 fmt:
-	npx prettier@'$(PRETTIER_VERSION)' "**/*.{yaml,yml,md}" --write
+	npx prettier@'$(PRETTIER_VERSION)' $(PRETTIER_PATTERN) --write

```